### PR TITLE
Transparent proxy support for cross-partition

### PR DESCRIPTION
**Pull request by freddygv**
There are a few changes here:
- Transparent proxy listener configuration needs to account for the new tagged address with virtual IPs generated by Consul. These should be considered both for cross-partition and partition-local traffic. The plain "virtual" tagged address should only be used for partition-local traffic, since the address is not guaranteed to be unique across partitions.
- The existing way of computing upstreams/downstreams from intentions was based on watching the "services" table and then evaluating a service's intentions against the unique service names. This won't work in 1.11 because the services table does not have a wildcard partition index, and cross-partition upstreams from intentions are supported. This watch was replaced with a watch of a table that stores service names by kind across all partitions.

### CODE REVIEW

File: agent/xds/listeners.go
Comment (markan) [inline agent/xds/listeners.go:184]: Why is this now > 2 ? Is this because there's a virtual IP and a tagged virtual IP?

**Review by kisunji — APPROVED** [review]

**Review by markan — APPROVED** [review]
This looks pretty reasonable to me. One question, but shouldn't block merge.

File: agent/xds/listeners.go
Comment (freddygv) [reply agent/xds/listeners.go:184]: Yep one is populated by consul-k8s and one by Consul, so if we detect more than two it's unexpected

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/517827.

**Pull request merged**
